plugin.video.invidious/resources/lib/invidious_plugin.py

```diff
@@ -0,0 +1,338 @@
+                self.add_directory_item(url=url, listitem=list_item, isFolder=True)
+
+        self.end_of_directory()
+
+    def display_new_search(self):
+        # query search with a dialog
+        dialog = xbmcgui.Dialog()
+        search_input = dialog.input(self.addon.getLocalizedString(30001), type=xbmcgui.INPUT_ALPHANUM)
+
+        self.display_search_result(search_input)
+
+    def display_search_result(self, search_input):
+        if len(search_input) == 0:
+            return
+
+        self.search_history.push(search_input)
+
+        xbmc.log(f"invidious searching for {search_input}.", xbmc.LOGDEBUG)
+
+        # pass search query to Invidious
+        results = self.api_client.search(search_input)
+
+        # assemble menu with the results
+        self.display_search_results(results)
+
+    def display_special_list(self, special_list_name):
+        if special_list_name not in self.__class__.SPECIAL_LISTS:
+            raise ValueError(str(special_list_name) + " is not a valid special list")
+
+        videos = self.api_client.fetch_special_list(special_list_name)
+
+        self.display_search_results(videos)
+
+    def display_channel_list(self, channel_id):
+        videos = self.api_client.fetch_channel_list(channel_id)
+
+        self.display_search_results(videos)
+
+    def display_playlist_list(self, playlist_id):
+        videos = self.api_client.fetch_playlist_list(playlist_id)
+
+        self.display_search_results(videos)
+
+    def play_video(self, id):
+        # TODO: add support for adaptive streaming
+        video_info = self.api_client.fetch_video_information(id)
+
+        xbmc.log(f"invidious playing video {video_info}.", xbmc.LOGDEBUG)
+
+        listitem = None
+        # check if playback via MPEG-DASH is possible
+        if not self.disable_dash and "dashUrl" in video_info:
+            is_helper = inputstreamhelper.Helper("mpd")
+            
+            if is_helper.check_inputstream():
+                url = video_info["dashUrl"]
+                xbmc.log(f"invidious using mpeg-dash stream {url}.", xbmc.LOGDEBUG)
+                listitem = xbmcgui.ListItem(path=url)
+                listitem.setProperty("inputstream", is_helper.inputstream_addon)
+                listitem.setProperty("inputstream.adaptive.manifest_type", "mpd")
+            else:
+                xbmc.log("invidious mpeg-dash input helper not available.", xbmc.LOGDEBUG)
+
+        # as a fallback, we use the last oldschool stream, as it is
+        # often the best quality.
+        if listitem is None:
+            url = video_info["formatStreams"][-1]["url"]
+            xbmc.log(f"invidious playback failing back to non-dash stream {url}!", xbmc.LOGINFO)
+            # it's pretty complicated to play a video by its URL in Kodi...
+            listitem = xbmcgui.ListItem(path=url)
+
+        datestr = datetime.utcfromtimestamp(video_info["published"]).date().isoformat()
+        listitem.setInfo('video', {
```

Comment:
setInfo is deprecated, please migrate to the new InfoTagVideo method. If you are not familiar thread below may help: https://forum.kodi.tv/showthread.php?tid=369255